### PR TITLE
Plotter

### PR DESCRIPTION
**Pull request by cdtn**
- [x] Implement concept of "displayed" attributes in `SeismicGeometry`.

- [x] Upgrade plotter functionality.

- [x] Fix instantaneous phase shift

- [x] Add tutorials

### CODE REVIEW

File: seismiqb/src/plotters.py [hunk 1]
```diff
@@ -46,11 +86,15 @@ def channelize_image(image, total_channels, color=None, greyscale=False, opacity
...
-    """ Filter the dict of kwargs leaving only supplied keys.
-    """
-    return {key: kwargs[key] for key in keys if key in kwargs}
-
+def filter_kwargs(kwargs, keys, condition_key=None, index=None, prefix=''):
+    """ Filter the dict of kwargs leaving only supplied keys. """
+    result = {}
+    if kwargs.get(condition_key, True):
+        for dst_key in keys:
+            src_key = prefix + dst_key
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:94]: If `title_fontsize` is None, you should use the `fontsize` key

Comment (SergeyTsimfer) [reply seismiqb/src/plotters.py:94]: That would significantly improve the code

File: seismiqb/src/plotters.py [hunk 2]
```diff
@@ -479,15 +534,97 @@ def overlap(cls, images, **kwargs):
...
+                legend_kwargs['color'] = layer_color
+                add_legend(ax, **legend_kwargs)
+        ax.set_title(**title_kwargs)
 
         return cls.save_and_show(fig, **updated)
 
+
+    @classmethod
+    def hist(cls, images, **kwargs):
+        """ Plot several images on one canvas using matplotlib: render the first one in greyscale
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:553]: Remove parameters from docstrings: they are unreachable anyways
Maybe, `multi_histogram`?

File: seismiqb/src/plotters.py [hunk 3]
```diff
@@ -479,15 +534,97 @@ def overlap(cls, images, **kwargs):
...
+
+
+        # form different groups of kwargs
+        figure_kwargs = filter_kwargs(updated, ['figsize', 'facecolor', 'dpi'])
+        title_kwargs = filter_kwargs(updated, ['label', 'fontsize', 'family', 'color', 'y'], prefix='title_')
+        xaxis_kwargs = filter_kwargs(updated, ['xlabel', 'fontsize', 'family', 'color'])
+        yaxis_kwargs = filter_kwargs(updated, ['ylabel', 'fontsize', 'family', 'color'])
+
+        # Create figure and axes
+        if 'ax' in kwargs:
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:606]: Move this logic of `create me multiple subplots or used supplied ` or `create me one axis or use supplied` into a separate methods and use it everywhere

**Review by SergeyTsimfer — CHANGES_REQUESTED** [review]

File: seismiqb/src/plotters.py [hunk 4]
```diff
@@ -46,10 +66,16 @@ def channelize_image(image, total_channels, color=None, greyscale=False, opacity
     return background
 
 
-def filter_kwargs(kwargs, keys):
-    """ Filter the dict of kwargs leaving only supplied keys.
-    """
-    return {key: kwargs[key] for key in keys if key in kwargs}
+def filter_kwargs(kwargs, keys, index=slice(None), prefix=''):
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:69]: Don't like the naming: you are not filtering `key-word arguments`, you are filtering supplied dictionary. `filter_parameters`?

File: seismiqb/src/plotters.py [hunk 5]
```diff
@@ -163,24 +210,27 @@ def single(cls, image, **kwargs):
         kwargs = cls.convert_kwargs('single', kwargs)
         # update defaults
         defaults = {'figsize': (12, 7),
-                    'cmap': 'viridis_r',
-                    'colorbar': True,
+                    'cmap': DEPTHS_CMAP,
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:213]:
```suggestion
                    'cmap': 'Depths',
```

File: seismiqb/src/plotters.py [hunk 6]
```diff
@@ -163,24 +210,27 @@ def single(cls, image, **kwargs):
...
+                    'colorbar_fraction': 0.5,
+                    'colorbar_aspect': 30}
         updated = {**defaults, **kwargs}
 
         # form different groups of kwargs
         figure_kwargs = filter_kwargs(updated, ['figsize', 'facecolor', 'dpi'])
         render_kwargs = filter_kwargs(updated, ['cmap', 'vmin', 'vmax', 'alpha', 'interpolation'])
-        label_kwargs = filter_kwargs(updated, ['label', 'y', 'fontsize', 'family', 'color'])
+        title_kwargs = filter_kwargs(updated, ['label', 'y', 'fontsize', 'family', 'color'], prefix='title_')
         xaxis_kwargs = filter_kwargs(updated, ['xlabel', 'fontsize', 'family', 'color'])
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:230]: Add `xaxis_` and `yaxis_` prefixes

File: seismiqb/src/plotters.py [hunk 7]
```diff
@@ -139,188 +292,171 @@ def single(cls, image, **kwargs):
...
+        if curve is not None:
+            curve = curve[offsets]
+            if curve.ndim == 1:
+                curve_x = (~np.isnan(curve)).nonzero()
+                curve_y = curve[curve_x]
+            # transform height-mask to heights if needed
+            elif curve.ndim == 2:
+                curve = curve[:, ::y_order]
+                curve = (~np.isnan(curve)).nonzero()
+                width = updated['horizon_width']
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:452]: `curve_width`?

File: seismiqb/src/horizon.py [hunk 8]
```diff
@@ -2366,8 +2368,8 @@ def __copy__(self):
         -------
         A horizon object with new matrix object and a reference to the old geometry attribute.
         """
-        return Horizon(np.copy(self.matrix), self.geometry, i_min=self.i_min, x_min=self.x_min,
-                       name=f'copy_of_{self.name}')
+        return type(self)(np.copy(self.matrix), self.geometry, i_min=self.i_min, x_min=self.x_min,
+                          name=f"copy_of_{self.name}")
```
[the remark this answers is not in the document]
Comment (cdtn) [reply seismiqb/src/horizon.py:2377]: dont

File: seismiqb/src/plotters.py [hunk 9]
```diff
@@ -109,10 +225,68 @@ def convert_kwargs(mode, kwargs):
...
+            colorbar_kwargs = filter_parameters(all_kwargs, cls.COLORBAR_KEYS, prefix='colorbar_', index=ax_num)
+            # if colorbar is disabled for subplot, add param to plot fake axis instead to keep proportions
+            colorbar_kwargs['fake'] = not colorbar_kwargs.pop('colorbar', True)
+            cls.add_colorbar(all_kwargs['axes_image'], **colorbar_kwargs)
+        if 'set_facecolor' in actions and all_kwargs.get('facecolor'):
+            ax.set_facecolor(all_kwargs['facecolor'])
+        if 'tick_params' in actions:
+            tick_params = filter_parameters(all_kwargs, cls.TICK_KEYS, index=ax_num)
+            ax.tick_params(**tick_params)
+        if 'disable_axes' in actions and all_kwargs.get('disable_axes'):
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:276]: Do you need this listed in `actions`?

File: seismiqb/src/plotters.py [hunk 10]
```diff
@@ -131,198 +305,213 @@ def save_and_show(fig, show=True, savepath=None, return_figure=False, pyqt=False
...
-        render_kwargs['cmap'] = cm
-
-        # Create figure and axes
-        if 'ax' in kwargs:
-            ax = kwargs['ax']
-            fig = ax.figure
-        else:
-            fig, ax = plt.subplots(**figure_kwargs)
+    @classmethod
+    def process_cmap(cls, params, bad_color):
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:319]: Not sure about this `inplace` update

File: seismiqb/src/plotters.py [hunk 11]
```diff
@@ -131,198 +305,213 @@ def save_and_show(fig, show=True, savepath=None, return_figure=False, pyqt=False
...
-        # channelize and plot the image
-        img = np.transpose(image.squeeze(), axes=updated['order_axes'])
-        xticks, yticks = updated.get('xticks', [0, img.shape[1]]), updated.get('yticks', [img.shape[0], 0])
-        extent = [xticks[0], xticks[-1], yticks[0], yticks[-1]]
+        try:
+            cmap = copy(plt.get_cmap(cmap_name))
+        except ValueError: # if not a valid cmap name, expect it to be a valid color name
+            if isinstance(cmap_name, str):
+                cmap_name = ColorConverter().to_rgb(cmap_name)
+            cmap = ListedColormap(cmap_name)
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:328]: Raise a `ValueError` if it is not a valid color either

File: seismiqb/src/plotters.py (hunk 10 above)
Comment (SergeyTsimfer) [reply seismiqb/src/plotters.py:319]: `cmap` is retrieved from `params`, while `bad_color` is passed explicitly. Seems inconsistent

File: seismiqb/src/plotters.py [hunk 12]
```diff
@@ -131,198 +305,213 @@ def save_and_show(fig, show=True, savepath=None, return_figure=False, pyqt=False
...
-
-        # Creating wiggle-curves and adding height-points if needed
-        xlim_curr = (0, len(image))
-        ylim_curr = (0, len(image[0]))
-        offsets = np.arange(*xlim_curr, xstep)
+                    'title': '',
+                    'label': '',
+                    # other
+                    'order_axes': (1, 0),
+                    'bad_color': (.0,.0,.0,.0)}
```
Comment (SergeyTsimfer) [inline seismiqb/src/plotters.py:407]: can you use `black`?

File: seismiqb/src/plotters.py (hunk 11 above)
Comment (cdtn) [reply seismiqb/src/plotters.py:328]: It's raised by `ColorConverter`.

**Review by SergeyTsimfer — APPROVED** [review]

File: seismiqb/src/plotters.py [hunk 13]
```diff
@@ -1,122 +1,316 @@
...
-        values in some cases.
-        """
-        # make conversion-dict for kwargs-keys
-        if mode in ['single', 'rgb', 'overlap', 'histogram', 'curve', 'histogram']:
-            keys_converter = {'title': 'label', 't':'label'}
-        elif mode == 'separate':
-            keys_converter = {'title': 't', 'label': 't'}
-        elif mode == 'grid':
-            keys_converter = {'title': 't'}
+    def nest_data(data, separate):
```
[the remark this answers is not in the document]
Comment (cdtn) [reply seismiqb/src/plotters.py:122]: `nest_data`

File: seismiqb/src/utility_classes.py [hunk 14]
```diff
@@ -309,9 +310,10 @@ def load_slide(cube_name, slide_no):
     All arguments to the decorated method must be hashable.
     """
     #pylint: disable=invalid-name, attribute-defined-outside-init
-    def __init__(self, maxsize=None, attributes=None, apply_by_default=True):
+    def __init__(self, maxsize=None, attributes=None, apply_by_default=True, copy_on_return=False):
```
Comment (SergeyTsimfer) [inline seismiqb/src/utility_classes.py:313]: `maybe_copy`?

**Review by SergeyTsimfer — APPROVED** [review]

**Review by SergeyTsimfer — APPROVED** [review]

**Pull request merged**